modules/core/src/main/java/org/apache/synapse/aspects/flow/statistics/opentracing/management/helpers/SpanTagger.java

```diff
@@ -40,37 +37,45 @@ private SpanTagger() {}
-            if (OpenTracingManagerHolder.isCollectingPayloads() || OpenTracingManagerHolder.isCollectingProperties()) {
-                StatisticDataUnit endEventDataUnit = (StatisticDataUnit) basicStatisticDataUnit;
-
-                if (OpenTracingManagerHolder.isCollectingPayloads()) {
-                    statisticsLog.setAfterPayload(endEventDataUnit.getPayload());
-                    span.setTag("beforePayload", statisticsLog.getBeforePayload());
-                    span.setTag("afterPayload", statisticsLog.getAfterPayload());
+        if (OpenTracingManagerHolder.isCollectingPayloads() || OpenTracingManagerHolder.isCollectingProperties()) {
+            if (OpenTracingManagerHolder.isCollectingPayloads()) {
+                span.setTag("beforePayload", openStatisticsLog.getBeforePayload());
+                if (spanWrapper.getCloseEventStatisticDataUnit() != null) {
+                    span.setTag("afterPayload", spanWrapper.getCloseEventStatisticDataUnit().getPayload());
+                } else {
+                    //This means a close event hasn't been triggered so payload is equal to before payload
+                    span.setTag("afterPayload", openStatisticsLog.getBeforePayload());
                 }
+            }
 
-                if (OpenTracingManagerHolder.isCollectingProperties()) {
-                    if (spanWrapper.getStatisticDataUnit().getContextPropertyMap() != null) {
-                        span.setTag("beforeContextPropertyMap",
-                                spanWrapper.getStatisticDataUnit().getContextPropertyMap().toString());
-                    }
-                    if (statisticsLog.getContextPropertyMap() != null) {
-                        span.setTag("afterContextPropertyMap", statisticsLog.getContextPropertyMap().toString());
+            if (OpenTracingManagerHolder.isCollectingProperties()) {
+                if (spanWrapper.getStatisticDataUnit().getContextPropertyMap() != null) {
+                    span.setTag("beforeContextPropertyMap",
+                            spanWrapper.getStatisticDataUnit().getContextPropertyMap().toString());
+                }
+                if (spanWrapper.getCloseEventStatisticDataUnit() != null) {
+                    if (spanWrapper.getCloseEventStatisticDataUnit().getContextPropertyMap() != null) {
+                        span.setTag("afterContextPropertyMap",
+                                spanWrapper.getCloseEventStatisticDataUnit().getContextPropertyMap().toString());
                     }
+                } else if (openStatisticsLog.getContextPropertyMap() != null) {
+                    span.setTag("afterContextPropertyMap", openStatisticsLog.getContextPropertyMap().toString());
+                }
+                if (spanWrapper.getCloseEventStatisticDataUnit() != null &&
+                        spanWrapper.getCloseEventStatisticDataUnit().getPropertyValue() != null) {
+                    span.setTag("propertyMediatorValue",
```

Comment:
Should not we need to add the property name with the value while setting the span tag?

Reply:
The property name is already being added